src/main/java/com/google/sps/data/Section.java

```diff
@@ -15,56 +15,101 @@
+  // with the first lesson being first in the list
+  private List<TimeRange> meetingTimes;
 
-    @Override
-    public boolean equals(Object other) {
-      if (other == this) {
-        return true;
-      }
+  public Section(String professor, List<TimeRange> meetingTimes) {
+    this.professor = professor;
+    this.meetingTimes = meetingTimes;
+    Collections.sort(this.meetingTimes, TimeRange.ORDER_BY_START);
 
-      if (!(other instanceof Section)) {
-        return false;
+    for (int i = 0; i < meetingTimes.size() - 1; i++) {
+      if (meetingTimes.get(i).overlaps(meetingTimes.get(i + 1))) {
+        throw new IllegalArgumentException("A section's lecture times can't overlap!");
       }
+    }
+  }
 
-      Section otherSection = (Section) other;
-
-      return otherSection.professor.equals(this.professor)
-              && otherSection.meetingTimes.equals(this.meetingTimes);
+  @Override
+  public boolean equals(Object other) {
+    if (other == this) {
+      return true;
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(professor, meetingTimes);
+    if (!(other instanceof Section)) {
+      return false;
     }
 
+    Section otherSection = (Section) other;
+
+    return otherSection.professor.equals(this.professor) 
+      && otherSection.meetingTimes.equals(this.meetingTimes);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(professor, meetingTimes);
+  }
+
+  public List<TimeRange> getMeetingTimes() {
+    return this.meetingTimes;
+  }
 
-    /**
-     * Function to return whether or not a section overlaps with another
-     * @param other
-     * @return If there's overlap between the current section and the other section
-     */
-    public boolean overlaps(Section other){
-        //TODO: This
-        return false;
+  /**
+   * Function to return whether or not a section overlaps with another.
+   * 
+   * @param other the section to check for overlap with.
+   * @return If there's overlap between the current section and the other section
+   */
+  public boolean overlaps(Section other) {
+    LinkedList<TimeRange> thisTimes = new LinkedList<>(this.meetingTimes);
+    LinkedList<TimeRange> otherTimes = new LinkedList<>(other.getMeetingTimes());
+
+    TimeRange thisCurr = thisTimes.removeFirst();
+    TimeRange otherCurr = otherTimes.removeFirst();
+    TimeRange thisFirst = thisCurr;
+    TimeRange otherFirst = otherCurr;
+
+    while (!thisTimes.isEmpty() || !otherTimes.isEmpty()) {
+      if (thisCurr.overlaps(otherCurr)) {
+        return true;
+      } else if (thisCurr.end() < otherCurr.end()) {
+        if (!thisTimes.isEmpty()) {
+          thisCurr = thisTimes.removeFirst();
+        } else {
+          otherCurr = otherTimes.removeFirst();
+        }
+      } else {
+        if (!otherTimes.isEmpty()) {
+          otherCurr = otherTimes.removeFirst();
+        } else {
+          thisCurr = thisTimes.removeFirst();
+        }
+      }
```

Comment:
I was thinking about whether or not there was a more elegant way to write this out, and I think there might be. It relies on the assumption that a given section cannot have two `TimeRange` elements that overlap in its `meetingTimes`.

1. Create a combined `List<TimeRange>` that contains the `TimeRange` elements from both sections' `meetingTimes`
2. Sort that list by `TimeRange` starting time
3. Iterate through the list and check to see if any two successive entries overlap
4. Add a check for the first & last elements in the list

It might look something like this instead, if you wanted to try something a little different using the `Stream` API... (note, that's more for the sake of demonstration/illustration, you and the pod can decide if you prefer this type of style versus an imperative loop - my bias is clear, but that's all is it; bias :)

```
ArrayList<TimeRange> combinedMeetingTimes = (....);
Collections.sort(combinedMeetingTimes, ....);
int meetingCount = combinedMeetingTimes.size();

return !IntStream.range(0, meetingCount + 1)
  .filter(index -> {
    TimeRange first = combinedMeetingTimes.get(index);
    TimeRange second = combinedMeetingTimes.get((index + 1) % meetingCount));
    return first.overlaps(second);
  })
  .findAny()
  .empty();
```

That's pretty close to the imperative version, too. It's a bit code golf-ey, could be made more legible with intermediate results getting assigned to name variables, but that's the idea in a nutshell.

For what it's worth, this approach handles the empty `TimeRange` list case for free.

Reply:
This approach is super elegant, but  I'm going to go with the imperative version. The issue is that I still need to compare the first and last elements, but only want to actually check those if the combined list is of size > 1, which requires manually handling the empty case before other logic is applied. (This is specific to the Saturday ->  Sunday edge case)

Reply:
I'm definitely going to switch the code over to something more similar to this suggestion though, because event though it's technically "slower" (O(nlogn) due to sort vs O(2n) in the original implementation), it's far more legible.

Reply:
> The issue is that I still need to compare the first and last elements, but only want to actually check those if the combined list is of size > 1, which requires manually handling the empty case before other logic is applied.

The code above performs this check. The imperative version would check for that as well, assuming it's structured the same. This is true because the code runs N iterations, where N = size of the list. When it gets to the last iteration, you'll have `first` be the last element in the list and `second` be the first element in the list.

Reply:
As for the running time, you're right that this is slower. The merging of the two already sorted lists isn't particularly smart in my example. If you really wanted to, you could write a O(n) merge algorithm and retain the O(n) overall runtime but I'm not convinced that's necessary given the use case here. Seems a tad overkill.

Reply:
I've gone back and had a look at your Sunday > Monday edge case handling. It seems less like an edge case you should be dealing with here and more of an edge case `TimeRange` should handle for you out of the box when you call its `overlaps()` method.

```
TimeRange a = Sunday at 11pm to Monday at 6am
TimeRange b = Monday at 4am to Monday at 8am

a.overlaps(b) => true
```

This is how it should be, no? It seems like the current implementation would return false here instead when in fact, these two `TimeRange` objects do represent ranges of time that overlap.

This comes down to the `TimeRange::contains()` implementation, which will return false if you ask it whether the `TimeRange a` from above contains the moment in time on Monday at 1am. I think addressing the problem there will save you some pain and prevent that edge case from escaping and running all around your codebase.

Reply:
What happens in the case of a single element list? I.E. one of the lists are empty, and the other has a single element.

Reply:
Fair point. This would pair up the element with itself since `index == (index + 1) % meetingCount` when `index = 0` and `meetingCount = 1`, which means that in this case it will find an overlap that isn't really there. Guess this can't escape having to check for the edge case where one or both lists are empty after all, not without some overly complicated or hacky shenanigans anyways. The easiest fix for the example code above then would be to throw a

```
if (meetingCount < 2) {
  return false;
}
```

in there before the stream.

Reply:
Addressed these! Ended up going with imperative because it felt a bit more idiomatic/legible. I still like the functional method though.